technical-reference/reference-technical-articles/technical-doc-diamond.md

```diff
@@ -0,0 +1,80 @@
+
+2. Each contract that needs to share functionality with the proxy contract is referred to as a "facet." Facets are separate contracts that can be upgraded or modified independently of each other and the proxy contract.
+
+3. The proxy contract delegates function calls to the correct facet contract based on the function selector. This allows the proxy contract to act as a single entry point for all shared functionality.
+4. The Diamond Proxy pattern also supports multiple inheritances. This means that each facet can inherit from other contracts, including other facets. This allows for a high degree of code reuse and modular design.
+
+## Enabling Venus Integration with the existing Unitroller - Diamond and Transparent proxy combination
+
+**Step 1:** The first step involved the creation of a new contract that serves as the Diamond proxy. This contract contains a mapping to hold the addresses of the facet contracts and a fallback function that delegates the function call to the appropriate facet. 
+
+Chained delegate calls are used, meaning calldata is delegated from Unitroller to the Diamond contract, and then a chained delegate call is made to a specific facet based on the function selector. 
+
+<figure><img src="../../.gitbook/assets/Screenshot from 2023-09-04 15-48-55.png" alt=""><figcaption></figcaption></figure>
+
+The Diamond contract checks for the facet address in the **selectors-to-facet-address** mapping and then makes a delegate call to that facet address. Facets inherit the ComptrollerStorage (ComptrollerV12Storage) to access the states.
+
+**Step 2:** Division of the comptroller into multiple facets based on the separation of concerns.
+
+Following the same eternal storage pattern i.e. inheriting storage contract for storage slots this basically replaced the AppStorage of Diamond pattern.
+
+Comptroller storage contains the facet addresses, mapping of the function selectors to the facet address and mapping of all selectors to the facet address.
+
+## User interaction with new diamond pattern
+The user interaction remains consistent, wherein users will continue to interact with the Unitroller proxy contract. However, a significant modification has occurred in the way the implementation contract is invoked. Now, the Diamond proxy serves as the implementation contract for the Unitroller, and the Comptroller's implementation has been subdivided into multiple facets. These facets will function as the various components of the Diamond proxy.
+
+Following a user's interaction with the Unitroller, the delegateCall operation will be directed towards the Diamond proxy. Once the delegateCall reaches the Diamond proxy, it will analyze the function selector and determine the appropriate facet address to which the delegateCall will be forwarded.
```

Comment:
```suggestion
Following a user's interaction with the Unitroller, the `delegateCall` operation will be directed towards the Diamond proxy. Once the `delegateCall` reaches the Diamond proxy, it will analyze the function selector and determine the appropriate facet address to which the `delegateCall` will be forwarded.
```